[Bug]: aws_batch_job_definition node properties with eks/ecs causes nil pointer

### Terraform Core Version

1.9.3

### AWS Provider Version

5.60.0

### Affected Resource(s)

aws_batch_job_definition

### Expected Behavior

When specifying other node properties, other than container, on multi-node resources, a panic occurs. It should be possible to define NodeRangeProperties with ecsProperties or eksProperties, not just containerProperties.

### Actual Behavior

The provider crashes due to a nil pointer: https://github.com/hashicorp/terraform-provider-aws/blob/main/internal/service/batch/node_properties.go#L27-L32.

[Enhancement] Node range properties doesn't support ecsProperties or eksProperties.
https://github.com/hashicorp/terraform-provider-aws/blob/main/internal/service/batch/node_properties.go#L20-L23


### Relevant Error/Panic Output Snippet

```shell
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x18 pc=0x111ea603c]

goroutine 88 [running]:
github.com/hashicorp/terraform-provider-aws/internal/service/batch.(*containerProperties).Reduce(0x0)
        github.com/hashicorp/terraform-provider-aws/internal/service/batch/container_properties.go:22 +0x1c
github.com/hashicorp/terraform-provider-aws/internal/service/batch.(*nodeProperties).Reduce(0x14000145200?)
        github.com/hashicorp/terraform-provider-aws/internal/service/batch/node_properties.go:29 +0x5c
github.com/hashicorp/terraform-provider-aws/internal/service/batch.EquivalentNodePropertiesJSON({0x0, 0x0}, {0x140001450e0, 0x10f})
        github.com/hashicorp/terraform-provider-aws/internal/service/batch/node_properties.go:67 +0x124
github.com/hashicorp/terraform-provider-aws/internal/service/batch.ResourceJobDefinition.func4({0x14001158b70?, 0x140020a5e00?}, {0x0?, 0x14001192140?}, {0x140001450e0?, 0x1195392c0?}, 0x140009df080?)
        github.com/hashicorp/terraform-provider-aws/internal/service/batch/job_definition.go:98 +0x34
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.schemaMap.diff(0x14001158b70, {0x1194f67c8, 0x14002093470}, {0x1141b9114, 0xf}, 0x14001192140, 0x140009df000, {0x1195392c0, 0x140009df080}, 0x0)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/schema.go:1143 +0x280
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.schemaMap.Diff(0x14001158b70, {0x1194f67c8, 0x14002093470}, 0x14001ddc9c0, 0x14000a22eb0, 0x14000224840, {0x119245ac0, 0x14002fdc5b0}, 0x0)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/schema.go:678 +0x298
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).SimpleDiff(0x1194f6bf8?, {0x1194f67c8?, 0x14002093470?}, 0x14001ddc9c0, 0x140020934a0?, {0x119245ac0?, 0x14002fdc5b0?})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/resource.go:990 +0x9c
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).PlanResourceChange(0x14001d69110, {0x1194f67c8?, 0x14002093380?}, 0x14000a22780)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/grpc_provider.go:858 +0xa30
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.(*muxServer).PlanResourceChange(0x140015505b0, {0x1194f67c8?, 0x140020930b0?}, 0x14000a22780)
        github.com/hashicorp/terraform-plugin-mux@v0.16.0/tf5muxserver/mux_server_PlanResourceChange.go:73 +0x250
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).PlanResourceChange(0x1400060c960, {0x1194f67c8?, 0x140020926c0?}, 0x140009de580)
        github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov5/tf5server/server.go:825 +0x2c8
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_PlanResourceChange_Handler({0x1190330e0, 0x1400060c960}, {0x1194f67c8, 0x140020926c0}, 0x140009de500, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:500 +0x1c0
google.golang.org/grpc.(*Server).processUnaryRPC(0x1400119c600, {0x1194f67c8, 0x14002092630}, {0x119558980, 0x1400052ef00}, 0x1400208cd80, 0x14004086870, 0x1229814e0, 0x0)
        google.golang.org/grpc@v1.63.2/server.go:1369 +0xb58
google.golang.org/grpc.(*Server).handleStream(0x1400119c600, {0x119558980, 0x1400052ef00}, 0x1400208cd80)
        google.golang.org/grpc@v1.63.2/server.go:1780 +0xb20
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.63.2/server.go:1019 +0x8c
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 16
        google.golang.org/grpc@v1.63.2/server.go:1030 +0x13c

Error: The terraform-provider-aws_v5.60.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```


### Terraform Configuration Files

An example with eksProperties.
```hcl
resource "aws_batch_job_definition" "test" {
  name = "tf_test_batch_job_definition_multinode"
  type = "multinode"
  retry_strategy {
    attempts = 1
  }

  node_properties = jsonencode({
    mainNode = 0
    nodeRangeProperties = [
      {
        targetNodes = "0:"
        eksProperties = {
          podProperties = {
            containers = [{
              image = "public.ecr.aws/amazonlinux/amazonlinux:1"
              command = [
                "sleep",
                "60"
              ]
              resources = {
                limits = {
                  cpu    = "1"
                  memory = "1024Mi"
                }
              }
            }]
          }
        }
      },
    ]
    numNodes = 1
  })
}
```

### Steps to Reproduce

Apply a manifest with a node property other than container.

### Debug Output

_No response_

### Panic Output

_No response_

### Important Factoids

_No response_

### References

https://docs.aws.amazon.com/batch/latest/APIReference/API_NodeRangeProperty.html
https://aws.amazon.com/about-aws/whats-new/2024/07/aws-batch-gang-scheduling-eks-multi-node-parallel-jobs/

### Would you like to implement a fix?

Yes

--- a/.changelog/40172.txt
+++ b/.changelog/40172.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_batch_job_definition: Fix crash when specifying `eksProperties` or `ecsProperties` block
+```

--- a/internal/service/batch/eks_properties.go
+++ b/internal/service/batch/eks_properties.go
@@ -1,0 +1,166 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package batch
+
+import (
+	"cmp"
+	"slices"
+	_ "unsafe" // Required for go:linkname
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	_ "github.com/aws/aws-sdk-go-v2/service/batch" // Required for go:linkname
+	awstypes "github.com/aws/aws-sdk-go-v2/service/batch/types"
+	smithyjson "github.com/aws/smithy-go/encoding/json"
+	tfjson "github.com/hashicorp/terraform-provider-aws/internal/json"
+	tfslices "github.com/hashicorp/terraform-provider-aws/internal/slices"
+)
+
+type eksProperties awstypes.EksProperties
+
+func (ep *eksProperties) reduce() {
+	if ep.PodProperties == nil {
+		return
+	}
+	ep.orderContainers()
+	ep.orderEnvironmentVariables()
+
+	// Set all empty slices to nil.
+	if len(ep.PodProperties.Containers) == 0 {
+		ep.PodProperties.Containers = nil
+	} else {
+		for j, container := range ep.PodProperties.Containers {
+			if len(container.Args) == 0 {
+				container.Args = nil
+			}
+			if len(container.Command) == 0 {
+				container.Command = nil
+			}
+			if len(container.Env) == 0 {
+				container.Env = nil
+			}
+			if len(container.VolumeMounts) == 0 {
+				container.VolumeMounts = nil
+			}
+			ep.PodProperties.Containers[j] = container
+		}
+	}
+	if len(ep.PodProperties.InitContainers) == 0 {
+		ep.PodProperties.InitContainers = nil
+	} else {
+		for j, container := range ep.PodProperties.InitContainers {
+			if len(container.Args) == 0 {
+				container.Args = nil
+			}
+			if len(container.Command) == 0 {
+				container.Command = nil
+			}
+			if len(container.Env) == 0 {
+				container.Env = nil
+			}
+			if len(container.VolumeMounts) == 0 {
+				container.VolumeMounts = nil
+			}
+			ep.PodProperties.InitContainers[j] = container
+		}
+	}
+	if ep.PodProperties.DnsPolicy == nil {
+		ep.PodProperties.DnsPolicy = aws.String("ClusterFirst")
+	}
+	if ep.PodProperties.HostNetwork == nil {
+		ep.PodProperties.HostNetwork = aws.Bool(true)
+	}
+	if len(ep.PodProperties.Volumes) == 0 {
+		ep.PodProperties.Volumes = nil
+	}
+	if len(ep.PodProperties.ImagePullSecrets) == 0 {
+		ep.PodProperties.ImagePullSecrets = nil
+	}
+}
+
+func (ep *eksProperties) orderContainers() {
+	slices.SortFunc(ep.PodProperties.Containers, func(a, b awstypes.EksContainer) int {
+		return cmp.Compare(aws.ToString(a.Name), aws.ToString(b.Name))
+	})
+}
+
+func (ep *eksProperties) orderEnvironmentVariables() {
+	for j, container := range ep.PodProperties.Containers {
+		// Remove environment variables with empty values.
+		container.Env = tfslices.Filter(container.Env, func(kvp awstypes.EksContainerEnvironmentVariable) bool {
+			return aws.ToString(kvp.Value) != ""
+		})
+
+		slices.SortFunc(container.Env, func(a, b awstypes.EksContainerEnvironmentVariable) int {
+			return cmp.Compare(aws.ToString(a.Name), aws.ToString(b.Name))
+		})
+
+		ep.PodProperties.Containers[j].Env = container.Env
+	}
+
+}
+
+func equivalentEKSPropertiesJSON(str1, str2 string) (bool, error) {
+	if str1 == "" {
+		str1 = "{}"
+	}
+
+	if str2 == "" {
+		str2 = "{}"
+	}
+
+	var ep1 eksProperties
+	err := tfjson.DecodeFromString(str1, &ep1)
+	if err != nil {
+		return false, err
+	}
+	ep1.reduce()
+	b1, err := tfjson.EncodeToBytes(ep1)
+	if err != nil {
+		return false, err
+	}
+
+	var ep2 eksProperties
+	err = tfjson.DecodeFromString(str2, &ep2)
+	if err != nil {
+		return false, err
+	}
+	ep2.reduce()
+	b2, err := tfjson.EncodeToBytes(ep2)
+	if err != nil {
+		return false, err
+	}
+
+	return tfjson.EqualBytes(b1, b2), nil
+}
+
+func expandEKSProperties(tfString string) (*awstypes.EksProperties, error) {
+	apiObject := &awstypes.EksProperties{}
+
+	if err := tfjson.DecodeFromString(tfString, apiObject); err != nil {
+		return nil, err
+	}
+
+	return apiObject, nil
+}
+
+// Dirty hack to avoid any backwards compatibility issues with the AWS SDK for Go v2 migration.
+// Reach down into the SDK and use the same serialization function that the SDK uses.
+//
+//go:linkname serializeEKSPProperties github.com/aws/aws-sdk-go-v2/service/batch.awsRestjson1_serializeDocumentEksProperties
+func serializeEKSPProperties(v *awstypes.EksProperties, value smithyjson.Value) error
+
+func flattenEKSroperties(apiObject *awstypes.EksProperties) (string, error) {
+	if apiObject == nil {
+		return "", nil
+	}
+
+	jsonEncoder := smithyjson.NewEncoder()
+	err := serializeEKSPProperties(apiObject, jsonEncoder.Value)
+
+	if err != nil {
+		return "", err
+	}
+
+	return jsonEncoder.String(), nil
+}

--- a/internal/service/batch/eks_properties.go
+++ b/internal/service/batch/eks_properties.go
@@ -11,7 +11,6 @@ import (
 	"github.com/aws/aws-sdk-go-v2/aws"
 	_ "github.com/aws/aws-sdk-go-v2/service/batch" // Required for go:linkname
 	awstypes "github.com/aws/aws-sdk-go-v2/service/batch/types"
-	smithyjson "github.com/aws/smithy-go/encoding/json"
 	tfjson "github.com/hashicorp/terraform-provider-aws/internal/json"
 	tfslices "github.com/hashicorp/terraform-provider-aws/internal/slices"
 )
@@ -132,35 +131,4 @@ func equivalentEKSPropertiesJSON(str1, str2 string) (bool, error) {
 	}
 
 	return tfjson.EqualBytes(b1, b2), nil
-}
-
-func expandEKSProperties(tfString string) (*awstypes.EksProperties, error) {
-	apiObject := &awstypes.EksProperties{}
-
-	if err := tfjson.DecodeFromString(tfString, apiObject); err != nil {
-		return nil, err
-	}
-
-	return apiObject, nil
-}
-
-// Dirty hack to avoid any backwards compatibility issues with the AWS SDK for Go v2 migration.
-// Reach down into the SDK and use the same serialization function that the SDK uses.
-//
-//go:linkname serializeEKSPProperties github.com/aws/aws-sdk-go-v2/service/batch.awsRestjson1_serializeDocumentEksProperties
-func serializeEKSPProperties(v *awstypes.EksProperties, value smithyjson.Value) error
-
-func flattenEKSroperties(apiObject *awstypes.EksProperties) (string, error) {
-	if apiObject == nil {
-		return "", nil
-	}
-
-	jsonEncoder := smithyjson.NewEncoder()
-	err := serializeEKSPProperties(apiObject, jsonEncoder.Value)
-
-	if err != nil {
-		return "", err
-	}
-
-	return jsonEncoder.String(), nil
 }

--- a/internal/service/batch/eks_properties.go
+++ b/internal/service/batch/eks_properties.go
@@ -96,7 +96,6 @@ func (ep *eksProperties) orderEnvironmentVariables() {
 
 		ep.PodProperties.Containers[j].Env = container.Env
 	}
-
 }
 
 func equivalentEKSPropertiesJSON(str1, str2 string) (bool, error) {

--- a/internal/service/batch/eks_properties_test.go
+++ b/internal/service/batch/eks_properties_test.go
@@ -1,0 +1,221 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package batch_test
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/hashicorp/terraform-provider-aws/internal/acctest/jsoncmp"
+	tfbatch "github.com/hashicorp/terraform-provider-aws/internal/service/batch"
+)
+
+func TestEquivalentEKSPropertiesJSON(t *testing.T) {
+	t.Parallel()
+
+	testCases := map[string]struct {
+		apiJSON           string
+		configurationJSON string
+		wantEquivalent    bool
+		wantErr           bool
+	}{
+		"empty": {
+			apiJSON:           ``,
+			configurationJSON: ``,
+			wantEquivalent:    true,
+		},
+		"reordered containers": {
+			apiJSON: `
+		{
+		    "podProperties": {
+		        "containers": [
+		          {
+		            "name": "container1",
+		            "image": "my_ecr_image1"
+		          },
+		          {
+		            "name": "container2",
+		            "image": "my_ecr_image2"
+		          }
+		        ]
+		    }
+		  }
+					`,
+			configurationJSON: `
+		{
+		    "podProperties": {
+		        "containers": [
+		          {
+		            "name": "container2",
+		            "image": "my_ecr_image2"
+		          },
+		          {
+		            "name": "container1",
+		            "image": "my_ecr_image1"
+		          }
+		        ]
+		    }
+		  }
+					`,
+			wantEquivalent: true,
+		},
+		"reordered environment": {
+			apiJSON: `
+		{
+		  "podProperties": {
+		      "containers": [
+		        {
+		          "name": "container1",
+		          "image": "my_ecr_image1",
+		          "env": [
+		            {
+		              "name": "VARNAME1",
+		              "value": "VARVAL1"
+		            },
+		            {
+		              "name": "VARNAME2",
+		              "value": "VARVAL2"
+		            }
+		          ]
+		        },
+		        {
+		          "name": "container2",
+		          "image": "my_ecr_image2",
+		          "env": []
+		        }
+		      ]
+		  }
+		}
+					`,
+			configurationJSON: `
+		{
+		  "podProperties": {
+		      "containers": [
+		        {
+		          "name": "container1",
+		          "image": "my_ecr_image1",
+		          "env": [
+		            {
+		              "name": "VARNAME2",
+		              "value": "VARVAL2"
+		            },
+		            {
+		              "name": "VARNAME1",
+		              "value": "VARVAL1"
+		            }
+		          ]
+		        },
+		        {
+		          "name": "container2",
+		          "image": "my_ecr_image2"
+		        }
+		      ]
+		    }
+		}
+					`,
+			wantEquivalent: true,
+		},
+		"full": {
+			apiJSON: `
+    {
+      "podProperties": {
+        "containers": [
+          {
+            "command": ["sleep", "60"],
+            "env": [
+              {
+                "name": "test",
+                "value": "Environment Variable"
+              }
+            ],
+            "image": "public.ecr.aws/amazonlinux/amazonlinux:1",
+            "volumeMounts": [],
+            "name": "container_a",
+            "resources": {
+              "requests": {
+                "cpu": "1.0",
+                "memory": "2048"
+              }
+            }
+          },
+          {
+            "command": ["sleep", "360"],
+            "env": [],
+            "image": "public.ecr.aws/amazonlinux/amazonlinux:1",
+            "volumeMounts": [],
+            "name": "container_b",
+            "resources": {
+              "requests": {
+                "cpu": "1.0",
+                "memory": "2048"
+              }
+            }
+          }
+        ],
+        "volumes": []
+      }
+    }
+		      `,
+			configurationJSON: `
+		{
+      "podProperties": {
+        "containers": [
+          {
+            "command": ["sleep", "60"],
+            "env": [
+              {
+                "name": "test",
+                "value": "Environment Variable"
+              }
+            ],
+            "image": "public.ecr.aws/amazonlinux/amazonlinux:1",
+            "name": "container_a",
+            "resources": {
+              "requests": {
+                "cpu": "1.0",
+                "memory": "2048"
+              }
+            }
+          },
+          {
+            "command": ["sleep", "360"],
+            "image": "public.ecr.aws/amazonlinux/amazonlinux:1",
+            "name": "container_b",
+            "resources": {
+              "requests": {
+                "cpu": "1.0",
+                "memory": "2048"
+              }
+            }
+          }
+        ]
+      }
+    }
+		      `,
+			wantEquivalent: true,
+		},
+	}
+
+	for name, testCase := range testCases {
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			output, err := tfbatch.EquivalentEKSPropertiesJSON(testCase.configurationJSON, testCase.apiJSON)
+			if got, want := err != nil, testCase.wantErr; !cmp.Equal(got, want) {
+				t.Errorf("EquivalentEKSPropertiesJSON err %t, want %t", got, want)
+			}
+
+			if err == nil {
+				if got, want := output, testCase.wantEquivalent; !cmp.Equal(got, want) {
+					t.Errorf("EquivalentEKSPropertiesJSON equivalent %t, want %t", got, want)
+					if want {
+						if diff := jsoncmp.Diff(testCase.configurationJSON, testCase.apiJSON); diff != "" {
+							t.Errorf("unexpected diff (+wanted, -got): %s", diff)
+						}
+					}
+				}
+			}
+		})
+	}
+}

--- a/internal/service/batch/exports_test.go
+++ b/internal/service/batch/exports_test.go
@@ -12,6 +12,7 @@ var (
 
 	EquivalentContainerPropertiesJSON       = equivalentContainerPropertiesJSON
 	EquivalentECSPropertiesJSON             = equivalentECSPropertiesJSON
+	EquivalentEKSPropertiesJSON             = equivalentEKSPropertiesJSON
 	EquivalentNodePropertiesJSON            = equivalentNodePropertiesJSON
 	ExpandEC2ConfigurationsUpdate           = expandEC2ConfigurationsUpdate
 	ExpandLaunchTemplateSpecificationUpdate = expandLaunchTemplateSpecificationUpdate

--- a/internal/service/batch/job_definition.go
+++ b/internal/service/batch/job_definition.go
@@ -790,7 +790,9 @@ func resourceJobDefinitionCreate(ctx context.Context, d *schema.ResourceData, me
 			}
 
 			for _, node := range props.NodeRangeProperties {
-				diags = append(diags, removeEmptyEnvironmentVariables(node.Container.Environment, cty.GetAttrPath("node_properties"))...)
+				if node.Container != nil {
+					diags = append(diags, removeEmptyEnvironmentVariables(node.Container.Environment, cty.GetAttrPath("node_properties"))...)
+				}
 			}
 			input.NodeProperties = props
 		}

--- a/internal/service/batch/job_definition_test.go
+++ b/internal/service/batch/job_definition_test.go
@@ -775,6 +775,102 @@ func TestAccBatchJobDefinition_NodeProperties_advanced(t *testing.T) {
 	})
 }
 
+func TestAccBatchJobDefinition_NodeProperties_withEKS(t *testing.T) {
+	ctx := acctest.Context(t)
+	var jd awstypes.JobDefinition
+	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+	resourceName := "aws_batch_job_definition.test"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(ctx, t); testAccPreCheck(ctx, t) },
+		ErrorCheck:               acctest.ErrorCheck(t, names.BatchServiceID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckJobDefinitionDestroy(ctx),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccJobDefinitionConfig_nodePropertiesEKS(rName),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckJobDefinitionExists(ctx, resourceName, &jd),
+					acctest.MatchResourceAttrRegionalARN(resourceName, names.AttrARN, "batch", regexache.MustCompile(fmt.Sprintf(`job-definition/%s:\d+`, rName))),
+					acctest.CheckResourceAttrEquivalentJSON(resourceName, "node_properties", `{
+						"mainNode": 0,
+						"nodeRangeProperties": [
+							{
+							"eksProperties": {
+								"podProperties": {
+								"containers": [
+									{
+									"args": [],
+									"command": ["sleep", "60"],
+									"env": [],
+									"image": "public.ecr.aws/amazonlinux/amazonlinux = 2",
+									"name": "test-eks-container-1",
+									"resources": { "requests": { "memory": "1024Mi", "cpu": "1" } },
+									"securityContext": {
+										"privileged": true,
+										"readOnlyRootFilesystem": true,
+										"runAsGroup": 3000,
+										"runAsNonRoot": true,
+										"runAsUser": 1000
+									},
+									"volumeMounts": []
+									}
+								],
+								"imagePullSecrets": [],
+								"initContainers": [],
+								"volumes": []
+								}
+							},
+							"instanceTypes": [],
+							"targetNodes": "0:"
+							}
+						],
+						"numNodes": 1
+						}`),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+				ImportStateVerifyIgnore: []string{
+					"deregister_on_new_revision",
+				},
+			},
+		},
+	})
+}
+
+func TestAccBatchJobDefinition_NodeProperties_withECS(t *testing.T) {
+	ctx := acctest.Context(t)
+	var jd awstypes.JobDefinition
+	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+	resourceName := "aws_batch_job_definition.test"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(ctx, t); testAccPreCheck(ctx, t) },
+		ErrorCheck:               acctest.ErrorCheck(t, names.BatchServiceID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckJobDefinitionDestroy(ctx),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccJobDefinitionConfig_nodePropertiesECS(rName),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckJobDefinitionExists(ctx, resourceName, &jd),
+					acctest.MatchResourceAttrRegionalARN(resourceName, names.AttrARN, "batch", regexache.MustCompile(fmt.Sprintf(`job-definition/%s:\d+`, rName))),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+				ImportStateVerifyIgnore: []string{
+					"deregister_on_new_revision",
+				},
+			},
+		},
+	})
+}
 func TestAccBatchJobDefinition_EKSProperties_basic(t *testing.T) {
 	ctx := acctest.Context(t)
 	var jd awstypes.JobDefinition
@@ -1364,6 +1460,104 @@ resource "aws_batch_job_definition" "test" {
     ]
 }
 CONTAINER_PROPERTIES
+}
+`, rName)
+}
+
+func testAccJobDefinitionConfig_nodePropertiesEKS(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_batch_job_definition" "test" {
+  name = %[1]q
+  type = "multinode"
+  retry_strategy {
+    attempts = 1
+  }
+
+  node_properties = jsonencode({
+    mainNode = 0
+    numNodes = 1
+    nodeRangeProperties = [{
+      targetNodes = "0:"
+      eksProperties = {
+        podProperties = {
+          containers = [
+            {
+              name  = "test-eks-container-1"
+              image = "public.ecr.aws/amazonlinux/amazonlinux = 2"
+              command = [
+                "sleep",
+                "60"
+              ]
+              resources = {
+                requests = {
+                  memory = "1024Mi"
+                  cpu    = "1"
+                }
+              }
+              securityContext = {
+                "runAsUser"              = 1000
+                "runAsGroup"             = 3000
+                "privileged"             = true
+                "readOnlyRootFilesystem" = true
+                "runAsNonRoot"           = true
+              }
+            }
+          ]
+        }
+      }
+    }]
+  })
+}
+  `, rName)
+}
+
+func testAccJobDefinitionConfig_nodePropertiesECS(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_batch_job_definition" "test" {
+  name = %[1]q
+  type = "multinode"
+  retry_strategy {
+    attempts = 1
+  }
+
+  node_properties = jsonencode({
+    mainNode = 0
+    numNodes = 1
+    nodeRangeProperties = [{
+      targetNodes = "0:"
+      ecsProperties = {
+        taskProperties = [{
+          containers = [{
+            image      = "public.ecr.aws/amazonlinux/amazonlinux:1"
+            command    = ["sleep", "60"]
+            name       = "container_a"
+            privileged = false
+            resourceRequirements = [{
+              value = "1"
+              type  = "VCPU"
+            },
+            {
+                value = "2048"
+                type  = "MEMORY"
+            }]
+            },
+            {
+              image   = "public.ecr.aws/amazonlinux/amazonlinux:1"
+              command = ["sleep", "360"]
+              name    = "container_b"
+              resourceRequirements = [{
+                value = "1"
+                type  = "VCPU"
+                },
+                {
+                  value = "2048"
+                  type  = "MEMORY"
+              }]
+          }]
+        }]
+      }
+    }]
+  })
 }
 `, rName)
 }

--- a/internal/service/batch/job_definition_test.go
+++ b/internal/service/batch/job_definition_test.go
@@ -1535,8 +1535,8 @@ resource "aws_batch_job_definition" "test" {
             resourceRequirements = [{
               value = "1"
               type  = "VCPU"
-            },
-            {
+              },
+              {
                 value = "2048"
                 type  = "MEMORY"
             }]

--- a/internal/service/batch/job_definition_test.go
+++ b/internal/service/batch/job_definition_test.go
@@ -835,6 +835,7 @@ func TestAccBatchJobDefinition_NodeProperties_withEKS(t *testing.T) {
 				ImportStateVerify: true,
 				ImportStateVerifyIgnore: []string{
 					"deregister_on_new_revision",
+					"node_properties",
 				},
 			},
 		},

--- a/internal/service/batch/node_properties.go
+++ b/internal/service/batch/node_properties.go
@@ -15,18 +15,29 @@ import (
 type nodeProperties struct {
 	MainNode            *int64
 	NodeRangeProperties []*nodeRangeProperty
-	NumNodes            *int64
+
+	NumNodes *int64
 }
 
 type nodeRangeProperty struct {
-	Container   *containerProperties
-	TargetNodes *string
+	Container     *containerProperties
+	EcsProperties *ecsProperties
+	EKSProperties *eksProperties
+	TargetNodes   *string
 }
 
 func (np *nodeProperties) reduce() {
 	// Deal with Environment objects which may be re-ordered in the API.
 	for _, node := range np.NodeRangeProperties {
-		node.Container.reduce()
+		if node.Container != nil {
+			node.Container.reduce()
+		}
+		if node.EcsProperties != nil {
+			node.EcsProperties.reduce()
+		}
+		if node.EKSProperties != nil {
+			node.EKSProperties.reduce()
+		}
 	}
 }
 

--- a/internal/service/batch/node_properties_test.go
+++ b/internal/service/batch/node_properties_test.go
@@ -134,6 +134,68 @@ func TestEquivalentNodePropertiesJSON(t *testing.T) {
 `,
 			wantEquivalent: true,
 		},
+		"Single node ECS Properties with multiple containers": {
+			apiJSON: `
+{
+	"mainNode": 1,
+	"nodeRangeProperties": [
+		{
+			"ecsProperties": {
+				"taskProperties": [
+				{
+					"containers": [
+					{
+						"name": "container1",
+						"image": "my_ecr_image1"
+					},
+					{
+						"name": "container2",
+						"image": "my_ecr_image2"
+					}
+					]
+				}
+				]
+			},
+			"targetNodes": "0:",
+			"environment": [],
+			"mountPoints": []
+		}
+	],
+	"numNodes": 1
+}
+`,
+			configurationJSON: `
+{
+  "mainNode": 1,
+  "nodeRangeProperties": [
+    {
+      "ecsProperties": {
+        "taskProperties": [
+          {
+            "containers": [
+              {
+                "name": "container2",
+                "image": "my_ecr_image2"
+              },
+              {
+                "name": "container1",
+                "image": "my_ecr_image1"
+              }
+            ]
+          }
+        ]
+      },
+      "targetNodes": "0:",
+      "environment": [],
+      "mountPoints": []
+    }
+  ],
+  "numNodes": 1
+}
+
+`,
+			wantEquivalent: true,
+		},
 	}
 
 	for name, testCase := range testCases {

